client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -279,8 +292,52 @@ private void checkConnected(final Handler<AsyncResult<Void>> resultHandler) {
         if (isConnectedInternal()) {
             resultHandler.handle(Future.succeededFuture());
         } else {
-            resultHandler.handle(Future.failedFuture(
-                    new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+            if (connecting.get()) {
+                // connect attempt in progress - let its completion complete the resultHandler here
+                log.debug("connection attempt to server [{}:{}] in progress, connection check will be completed with its result",
+                        connectionFactory.getHost(), connectionFactory.getPort());
+                final Promise<Void> promiseToAdd = Promise.promise();
+                promiseToAdd.future().setHandler(resultHandler);
+                // atomically add the promise to checkConnectedPromises - but only if checkConnectedPromises hasn't been cleared already
+                final List<Promise<Void>> newCheckConnectedPromises = checkConnectedPromises
+                        .accumulateAndGet(Collections.singletonList(promiseToAdd), (existing, toAdd) -> {
+                            // no modification of the existing list done here, keeping the accumulatorFunction function side-effect free as required
+                            if (existing == null) {
+                                return null;
+                            }
+                            final List<Promise<Void>> promises = new ArrayList<>(existing.size() + 1);
+                            promises.addAll(existing);
+                            promises.add(toAdd.get(0));
+                            return promises;
+                        });
+                if (newCheckConnectedPromises == null) {
+                    // checkConnectedPromises wasn't updated with our promiseToAdd because the checkConnectedPromises were cleared in between
+                    resultHandler.handle(Future.failedFuture(
+                            new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+                }
+            } else {
+                resultHandler.handle(Future.failedFuture(
+                        new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+            }
+        }
+    }
+
+    private void removeAndCompleteCheckConnectedPromises(final AsyncResult<HonoConnection> connectionResult) {
+        final List<Promise<Void>> promises = checkConnectedPromises.getAndSet(null);
+        if (promises != null) {
+            if (connectionResult.succeeded()) {
+                promises.forEach(promise -> runOnContextOrDirectly(v -> promise.complete()));
+            } else {
+                promises.forEach(promise -> runOnContextOrDirectly(v -> promise.fail(connectionResult.cause())));
+            }
+        }
+    }
+
+    private void runOnContextOrDirectly(final Handler<Void> action) {
+        if (context != null) {
```

Comment:
FMPOV we can simply always do
```
vertx.getOrCreateContext().runOnContext(v -> action.handle(null));
```
I do not think that it is important, which event loop thread it is that the handler will be invoked on. If the connection has been established successfully, then any subsequent attempts to use that connection will always be executed on the connection's context anyway ...

Reply:
Ok, yes. I've amended the commit accordingly.